tutorials/docs-17-implementing-samplers/index.qmd

```diff
@@ -0,0 +1,479 @@
+```
+
+Yep, that still works, but only because Turing.jl actually *transforms* the `turing_model` from constrained to unconstrained, so that the `sampler` provided to `externalsampler` is actually always working in unconstrained space! This is not always desirable, so we can turn this off:
+
+```{julia}
+chain = sample(turing_model, externalsampler(sampler; unconstrained=false), 10_000; progress=false)
+```
+
+The fun thing is that this still sort of works because
+
+```{julia}
+logpdf(Beta(3, 3), 10.0)
+```
+
+and so the samples that fall outside of the range are always rejected. But do notice how much worse all the diagnostics are, e.g. `ess_tail` is very poor compared to when we use `unconstrained=true`. Moreover, in more complex cases this won't just result in a "nice" `-Inf` log-density value, but instead will error:
+
+```{julia}
+@model function demo()
+    σ² ~ truncated(Normal(), lower=0)
+    # If we end up with negative values for `σ²`, the `Normal` will error.
+    x ~ Normal(0, σ²)
+end
+sample(demo(), externalsampler(sampler; unconstrained=false), 10_000; progress=false)
+```
+
+As expected, we run into a `DomainError` at some point, while if we set `unconstrained=true`, letting Turing.jl transform the model to a unconstrained form behind the scenes, everything works as expected:
+
+```{julia}
+sample(demo(), externalsampler(sampler; unconstrained=true), 10_000; progress=false)
+```
+
+Neat!
+
+Similarly, which automatic differentiation backend one should use can be specified through the `adtype` keyword argument too. For example, if we want to use [ReverseDiff.jl](https://github.com/JuliaDiff/ReverseDiff.jl) instead of the default [ReverseDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl):
+
+```{julia}
+using ReverseDiff: ReverseDiff
+# Specify that we want to use `AutoReverseDiff`.
+sample(
+    demo(),
+    externalsampler(sampler; unconstrained=true, adtype=AutoReverseDiff()),
+    10_000;
+    progress=false
+)
+```
+
+Double-neat.
+
+## Summary
+
+At this point it's worth maybe reminding ourselves what we did and also *why* we did it:
+
+1.  We define our models in the [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) interface because it makes the sampler agnostic to how the underlying model is implemented.
+2.  We implement our sampler in the [AbstractMCMC.jl](https://github.com/TuringLang/AbstractMCMC.jl) interface, which just means that our sampler is a subtype of `AbstractMCMC.AbstractSampler` and we implement the MCMC transition in `AbstractMCMC.step`.
+3.  (1) and (2) makes it so our sampler can be used with a wide range of model implementations, amongst them being models implemented in both Turing.jl and Stan. This gives you, the inference implementer, a large collection of models to test your inference method on, in addition to allowing users of Turing.jl and Stan to try out your inference method with minimal effort.
```

Comment:
Quarto renders the `(1) and (2)` as "1. and (2)" with a funny indent. I think technically the `(1)` starts another nested enumerated list.